testsuite/src/it/java/org/graalvm/tests/integration/PerfCheckTest.java

```diff
@@ -552,6 +539,169 @@ public void testQuarkusFullMicroProfile(TestInfo testInfo) throws IOException, I
+                    report.put("executableSizeKb", String.valueOf(executableSizeKb));
+                    l = parseSerialGCLog(processLog.toPath(), statsFor, false);
+                    report.put("incrementalGCevents", String.valueOf(l.incrementalGCevents));
+                    report.put("fullGCevents", String.valueOf(l.fullGCevents));
+                } else {
+                    l = parseSerialGCLog(processLog.toPath(), statsFor, true);
+                    report.put("incrementalGCevents", "-1");
+                    report.put("fullGCevents", "-1");
+                    report.put("executableSizeKb", "-1");
+                }
+                report.put("timeSpentInGCs", String.valueOf(l.timeSpentInGCs));
+                report.put("testMethod", cn + "#" + mn);
+                reports.add(report);
+            }
+
+            // log the report
+            final String reportPayload = mapToJSON(reports);
+            LOGGER.info(reportPayload);
+
+            LOGGER.info("Wait till the ports close...");
+            assertTrue(waitForTcpClosed("localhost", parsePort(app.urlContent.urlContent[0][0]), 60),
+                    "Main is still open.");
+            Logs.checkLog(cn, mn, app, processLog);
+
+            // TODO, need to implement some asserting here once the timeSpentInGCs value is not 0
+        } finally {
+            // final cleanup after the test is over
+            if (process != null) {
+                processStopper(process, true);
+            }
+            Logs.archiveLog(cn, mn, Path.of(appDir.getAbsolutePath(),
+                    "target", "quarkus-native-image-source-jar", "quarkus-json.json").toFile());
+            Logs.archiveLog(cn, mn, processLog);
+            cleanTarget(app);
+            if (patch != null) {
+                runCommand(getRunCommand("git", "apply", "-R", patch), appDir);
+            }
+        }
+    }
+
+    private void generateRequestsWithHyperfoil(Apps app, File appDir, File processLog, String cn, String mn)
+            throws IOException, InterruptedException, URISyntaxException {
+        try {
+            removeContainer("hyperfoil-container");
+
+            // start Hyperfoil
+            final List<String> getAndStartHyperfoil = getRunCommand(app.buildAndRunCmds.runCommands[2]);
+            Process hyperfoilProcess = runCommand(getAndStartHyperfoil, appDir, processLog, app);
+            assertNotNull(hyperfoilProcess, "Hyperfoil failed to run. Check " + getLogsDir(cn, mn) + File.separator + processLog.getName());
+            LOGGER.info("Hyperfoil process started with pid " + hyperfoilProcess.pid());
+            Commands.waitForContainerLogToMatch(ContainerNames.HYPERFOIL.name,
+                    Pattern.compile(".*Hyperfoil controller listening.*", Pattern.DOTALL), 600, 5, TimeUnit.SECONDS);
+            WebpageTester.testWeb(app.urlContent.urlContent[1][0], 15, app.urlContent.urlContent[1][1], false);
+
+            // upload the benchmark
```

Comment:
@patrikcerbak Since "Upload Benchmark to Hyperfoil" is likely excatly the same with JFR, would you mind factoring it out to a stateless utility method more tests could call?